### DISCUSSION TITLE
Java application exits when nodeRuntime.resetContext() executes

### DISCUSSION THREAD

**Discussion by mrinalsharma** [caoccao/Javet#104] [Q&A]
In my java application, I am using Javet to run a NodeJS HTTP server and some other NodeJS scripts. When I try to reset context with noderuntime.resetContext(), I notice that my entire application exits. This issue doesn't happen if I run a very basic NodeJs script without HTTP server code. At the point when resetContext is called, I also see the following error in the console logs.

```
Error: listen EADDRINUSE: address already in use 127.0.0.1:8000
    at Server.setupListenHandle [as _listen2] (net.js:1320:16)
    at listenInCluster (net.js:1368:12)
    at GetAddrInfoReqWrap.doListen [as callback] (net.js:1505:7)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:71:8)
Emitted 'error' event on Server instance at:
    at emitErrorNT (net.js:1347:8)
    at processTicksAndRejections (internal/process/task_queues.js:82:21) {
  code: 'EADDRINUSE',
  errno: -4091,
  syscall: 'listen',
  address: '127.0.0.1',
  port: 8000
}
```

I have reviewed https://www.caoccao.com/Javet/tutorial/advanced/interact_with_node_js.html before working on a solution.

**Comment by caoccao** [comment]
Node.js has the capability of terminating the whole process (in this case, the JVM) when it encounters an unrecoverable error. I think your case is not different from running your code directly in Node.js.

Here are what I would do for your reference.

1. Try `resetIsolate()` instead.
2. Force the JS server to quit completely so that the port is released.
3. Post the minimal reproducible code (JS + Java) for troubleshooting.

Please let me know if these make sense.

**Reply by mrinalsharma** [reply]
I tried option 2 "Force the JS server to quit completely so that the port is released." and it worked.

**Reply by mrinalsharma** [reply]
For forcing the JS code to close the Server before noderuntime is reset, I defined a method named Stop() in JS . This method returns a promise.
```
   return new Promise((resolve, reject) => {
      server.close((err) => {
        resolve("SUCCESS");
      })
    })
```

In the Java code, I wait until the promise is pending.
```
V8ValuePromise v8ValuePromise = comp.invoke("stop");
while(v8ValuePromise.getState() == v8ValuePromise.STATE_PENDING) {
	System.out.println("Promise State: "+ v8ValuePromise.getState());
}
nodeRuntime.resetContext();
iJavetEngine.close();
```

I hope this helps others too.

**Reply by caoccao** [reply]
Thank you for the sharing. That's very helpful to the others.

## 2 Recommendations

1. It's better to apply `try-with-resource` to `v8ValuePromise`.
2. Sleeping a while in the `while` block would help easy the CPU usage.

**Comment by mrinalsharma** [comment]
In the application, I am giving an ability to the user of the application to write their own NodeJS script. What would be the right way to decide if nodeRuntime.await(); need to be called and when?

E.g. in the case of HTTP server ```nodeRuntime.await();``` should be called after execution of the below code in NodeJS script.
```  
server.listen(port, host, () => {
      console.log(`Server is running on http://${host}:${port}`);
    })
```

**Reply by caoccao** [reply]
I suggest you register an API e.g. `myAwait()` for the script to invoke explicitly. That API calls `NodeRuntime.await()` internally. I think giving the control to the user is not a bad idea in your use case. Hope it makes sense.

**Reply by mrinalsharma** [reply]
Noderuntime.await() is a blocking call, will it not cause JavaScript code to halt execution.

Also if there are two servers started in a script. When should await() be called? After first listen or second?

**Reply by caoccao** [reply]
> Noderuntime.await() is a blocking call, will it not cause JavaScript code to halt execution.

It's blocking from Java perspective. It's non-blocking from JS perspective because it releases the event loop periodically. 

> Also if there are two servers started in a script. When should await() be called? After first listen or second?

As the doc shows, you are recommended to host one JS server per Java thread. So, both of them can work. Of course, the server ports need to be different.

This is quite new to Node.js developers because Javet allows multiple Node.js instances running together.
